lib/grizzly/zwave_firmware.ex

```diff
@@ -0,0 +1,167 @@
+defmodule Grizzly.ZwaveFirmware do
+  @moduledoc """
+  Controller ZWave firmware update support
+  """
+  require Logger
+
+  @doc """
+  Update the ZWave chip firmware if zw_programmer is present and a
+  valid GBL file is available that would up the firmware version.
+  """
+  @spec maybe_run_zwave_firmware_update(Grizzly.Options.t()) :: :ok
+  def maybe_run_zwave_firmware_update(options) do
+    with :ok <- check_firmware_dir(options),
+         :ok <- check_zwave_programmer_exists(options),
+         {:ok, gbl_file} <- get_firmware_file(options) do
+      Logger.info("[Grizzly]: Trying to upload firmware image #{inspect(gbl_file)}")
+      _ = maybe_upload_zwave_firmware_update(gbl_file, options)
+    else
+      {:error, :no_firmware_dir} ->
+        _ =
+          Logger.info(
+            "[Grizzly]: no firmware directory configured, not attempting a firmware update"
+          )
+
+      {:error, :firmware_dir_does_not_exist} ->
+        Logger.info(
+          "[Grizzly]: firmware directory does not exist, not attempting a firmware update"
+        )
+
+      {:error, :zw_programmer_does_not_exist} ->
+        Logger.info(
+          "[Grizzly]: zw_programmer binary does not exist, not attempting a firmware update"
+        )
+
+      {:error, :no_zwave_firmware_file} ->
+        Logger.info("[Grizzly]: could not detect firmware file, not attempting a firmware update")
+    end
+
+    :ok
+  end
+
+  defp check_firmware_dir(%{zwave_firmware_dir: nil}) do
+    {:error, :no_firmware_dir}
+  end
+
+  defp check_firmware_dir(options) do
+    Logger.info("[Grizzly] Checking for firmware dir #{inspect(options.zwave_firmware_dir)}")
+
+    if File.exists?(firmware_dir(options)) do
+      :ok
+    else
+      {:error, :firmware_dir_does_not_exist}
+    end
+  end
+
+  defp firmware_dir(options), do: Application.app_dir(:grizzly, options.zwave_firmware_dir)
+
+  defp check_zwave_programmer_exists(options) do
+    if File.exists?(options.zw_programmer_path) do
+      :ok
+    else
+      {:error, :zw_programmer_does_not_exist}
+    end
+  end
+
+  defp get_firmware_file(options) do
+    dir = firmware_dir(options)
+
+    case File.ls!(dir) do
```

Comment:
This function should either raise or return error tuples if the file isn't found. I was going to suggest changing `File.ls!/1` to `File.ls/1`, but I think this needs to be more explicit on what file is loaded. The restriction of having a directory with only one file in it seems limiting. It seems easy to imaging having a couple firmware files for different Z-Wave modules - like a 500 series one, a 700 series one and an 800 series one.

Reply:
I think what @fhunleth is saying is that for different chip types (500, 700, 800) we might be able to use the same Z-Wave firmware (for example `.hex` files for 500 vs `.gbl` for 700).

I am not sure the best way to go about having different zwave firmware for different chips and maybe we can look at this type of support later? I have some thoughts, but if we can get a single firmware loading up automatically and test that out for now I think that moves forward in a useable feature?

One thing I think we should check against in regards to different chip versions is the file extension vs the chip type.

When `zw_progammer` tells us the SDK version it will tell us the chip type. Where the pattern is `n = n00` so `7 = 700`. 

For example:


```
if  chip type == 5 and file extension == .hex, then ok

if chip type == 7 and file extension == .gbl, then ok

otherwise, log we don't recognize the chip type
```

The `zw_programmer` program does not protect sending an unsupported firmware format to a chip.

Reply:
I added code for this. We won't be able to use the same firmware for different modules - the 500-series and 700-series chips have microcontrollers. What we'll need to eventually do is decide at runtime which firmware to apply since devices might have 500-series parts on them or 700-series or 800-series depending on when they're made and what can be bought.